### ISSUE TITLE
Clarify clGetEventProfilingInfo error discrepancy

### ISSUE THREAD

**Issue by jlewis-austin** [KhronosGroup/OpenCL-Docs#493]
In the OpenCL 2.0-2.2 API specs, the CL_PROFILING_INFO_NOT_AVAILABLE error description includes the following phrase:
_or if event refers to the clEnqueueSVMFree command or is a user event object_

In 3.0.5. it is gone. Need to clarify if that was intentional, and either add it to the 3.0 spec or note the change.

**Comment by bashbaug** [comment]
Just to be clear, the part that is different is "if event refers to clEnqueueSVMFree", and the part about "or is a user event object" still exists in the latest OpenCL 3.0 specs.

If so, this was intentional.  This change was made as an OpenCL 2.2 spec bugfix between spec revisions v2.2-7 and v2.2-8, which is why it doesn't show up in any of the change logs.

For Khronos folks, this was discussed in internal issue 108.  A summary of the issue is "which commands should carry profiling information" and the resolution was effectively "that any event that results from enqueuing a command  should carry profiling information when profiling is enabled for the queue", hence there is no need for an exception for clEnqueueSVMFree (or any other enqueue API).

**Comment by jlewis-austin** [comment]
OK thanks. I'm not sure if that merits a mention in the 3.0 spec. Feel free to close if not.

**Comment by bashbaug** [comment]
Hi @jlewis-austin, what would you like to do for this issue?  Should we close it with no changes?  Or do you think we should add a note in the 3.0 spec?  If we should add a note, could you suggest where it should go and approximately what it should say?  Thanks!

**Comment by jlewis-austin** [comment]
@bashbaug the main question is what behavior will be expected of new 2.0-2.2 implementations going forward. If they stick to their current spec wording they will need some work to become 3.0 compliant. 
I could go either way (I know this since I changed my mind in the middle of answering), but I'm thinking for clarity it would be nice to have a note about it. I'll take a crack at it and if that doesn't work out we can drop it and let the CTS negative tests define expected behaviors for each version.

**Comment by jlewis-austin** [comment]
@bashbaug coming back to this finally, I'm thinking now that the current unified 3.0 spec requirement is clear and should be interpreted to apply to all new implementations of any version going forward. 

If we want to add clarification, though, I'd recommend a footnote on the CL_PROFILING_INFO_NOT_AVAILABLE description:
[x] OpenCL 2.x requires CL_PROFILING_INFO_NOT_AVAILABLE to be returned for events referring to the **clEnqueueSvmFree** command. This restriction is removed and new implementations should implement profiling information support for this command.

**Comment by bashbaug** [comment]
Good idea, I think adding to the CL_PROFILING_INFO_NOT_AVAILABLE description for [clGetEventProfilingInfo](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clGetEventProfilingInfo) is the right place to add this, but because this would be changing normative spec behavior I'm re-thinking whether it should be a footnote.

How about adding a sentence to the main spec error description like:

> ... or if event is a user event object.  Prior to OpenCL 3.0, implementations may return CL_PROFILING_INFO_NOT_AVAILABLE from clGetEventProfilingInfo for an event created by clEnqueueSvmFree.

Basically, this would document that we prefer that OpenCL 2.x implementations still follow the unified 3.0 spec behavior, while acknowledging that there still may be older OpenCL 2.x implementations that return an error in this case.

Note: I did a quick search through the CTS tests and there do not appear to be any tests for this case.

**Comment by jlewis-austin** [comment]
@bashbaug thanks, that works for me
